### PR TITLE
Update validation, remove slicing logic from classes

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/modules/transformer.py [hunk 1]
```diff
@@ -429,20 +429,13 @@ def setup_caches(
...
-        """
-        return self.encoder_max_cache_seq_len is not None
-
-    def decoder_caches_are_enabled(self) -> bool:
-        """Check if there are any :class:`~torchtune.modules.TransformerCrossAttentionLayer`
-        layers which have cache enabled."""
-        return self.decoder_max_cache_seq_len is not None
+    def caches_are_enabled(self) -> bool:
+        """Check if the key value caches are setup."""
+        return self.layers[0].cache_enabled
```
Comment (felipemello1) [inline torchtune/modules/transformer.py:434]: I see that in the old logic we checked encoder/decoder. It seems that this is not necessary, and we can always check layer 0

is it enough? It makes sense to me, just double checking

File: torchtune/modules/transformer.py [hunk 2]
```diff
@@ -429,20 +429,13 @@ def setup_caches(
...
-        or :class:`~torchtune.modules.fusion.FusionLayer` layers which have cache enabled.
-        """
-        return self.encoder_max_cache_seq_len is not None
-
-    def decoder_caches_are_enabled(self) -> bool:
-        """Check if there are any :class:`~torchtune.modules.TransformerCrossAttentionLayer`
-        layers which have cache enabled."""
-        return self.decoder_max_cache_seq_len is not None
+    def caches_are_enabled(self) -> bool:
+        """Check if the key value caches are setup."""
```
Comment (felipemello1) [inline torchtune/modules/transformer.py:433]: nit: fine as it is, but maybe add context about how/when this is used. Something like: "useful during inference to xyz". Feel free to ignore

File: torchtune/modules/transformer.py [hunk 3]
```diff
@@ -500,26 +495,23 @@ def _validate_inputs(
...
                 )
 
-        if self.encoder_caches_are_enabled():
-            if encoder_mask is None:
+            if encoder_input is None and encoder_mask:
                 raise ValueError(
-                    "KV-caches for cross-attention/fusion layers are setup for inference mode, causal masks must be provided!"
-                    " Use the `encoder_mask` arg to provide a causal mask."
+                    "KV-caches for cross-attention/fusion layers are setup for inference mode and you seem to be using"
+                    " encoder_input, causal masks must be provided! Use the `encoder_mask` arg to provide a causal mask."
```
Comment (felipemello1) [inline torchtune/modules/transformer.py:508]: nit: prob a period instead of comma would be better

File: torchtune/modules/transformer.py [hunk 4]
```diff
@@ -500,26 +495,23 @@ def _validate_inputs(
...
+        if self.caches_are_enabled():
             if mask is None:
                 raise ValueError(
                     "KV-caches for self-attention layers are setup for inference mode, causal masks must be provided!"
                     " Use the `mask` arg to provide a causal mask."
                 )
 
-        if self.encoder_caches_are_enabled():
-            if encoder_mask is None:
+            if encoder_input is None and encoder_mask:
```
Comment (felipemello1) [inline torchtune/modules/transformer.py:505]: in there some good default for encoder_mask, like causal mask? If not, just mark as resolved

**Review by felipemello1 — COMMENTED** [review]
Minor nits. I worry about the logic for checking if cache is enabled. I dont think it is robust enough.

I would like to see a test confirming that these errors are properly raised. Do you think its possible?

File: torchtune/modules/transformer.py (hunk 4 above)
Comment (joecummings) [reply torchtune/modules/transformer.py:505]: I plan on removing / re-working most of this validation logic to have better defaults, but that will be a follow-up PR.

**Review by felipemello1 — APPROVED** [review]
stamping to unblock, but IMO adding a small test before merging would be much better

**Pull request merged**